### PR TITLE
Ft/move md backends from S3

### PR DESCRIPTION
**Pull request by JianqinWang**
Extract metadata backends from S3 to here so backbeat can use these features.
Merge this PR before https://github.com/scality/S3/pull/1129 and https://github.com/scality/backbeat/pull/187
JIRA ticket: https://scality.atlassian.net/browse/ZENKO-140

### CODE REVIEW

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 1]
```diff
@@ -70,63 +69,57 @@ function generatePHDVersion(versionId) {
  * @param {Object} params - constructor params
  * @param {String} params.host - host for mongo
  * @param {String} params.port - port for mongo
- * @param {String} params.writeConcern - writeConcern for mongo
- * @param {String} params.replicaSet - replicaSet for mongo
```
Comment (rahulreddy) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Need to retain `params.replicaSet, params.replicationGroupId ` as it's being passed as a param.

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 2]
```diff
@@ -70,63 +69,57 @@ function generatePHDVersion(versionId) {
...
+            host,
+            replicaSet);
         this.logger = logger;
         this.client = null;
         this.db = null;
         this.path = path;
         this.replicationGroupId = replicationGroupId;
         this.database = database;
-        this.logger.info(`connecting to ${mongoUrl}`);
+        this.logger.debug(`connecting to ${this.mongoUrl}`);
```
Comment (rahulreddy) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Please remove this log line as it's misleading. This is just a constructor, there is no connect method called.

File: lib/storage/metadata/mongoclient/MongoClientInterface.js [hunk 3]
```diff
@@ -178,9 +172,6 @@ class MongoClientInterface {
             _id: bucketName,
         }, {}, (err, doc) => {
             if (err) {
-                log.error(
```
Comment (rahulreddy) [inline lib/storage/metadata/mongoclient/MongoClientInterface.js]: Is there a reason why the error logs are being removed?

File: lib/storage/metadata/wrapper.js [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    // wrap private function in a try/catch clause
+    // just in case JSON parsing throws an exception
+    try {
+        return _parseListEntries(entries);
+    } catch (e) {
+        return e;
+    }
+}
+
+class Metadata {
```
Comment (rahulreddy) [inline lib/storage/metadata/wrapper.js]: This class name, export and filename has to be `MetadataWrapper`

File: lib/storage/metadata/mongoclient/MongoClientInterface.js (hunk 3 above)
Comment (JianqinWang) [reply lib/storage/metadata/mongoclient/MongoClientInterface.js]: this branch was pulled off an older version of Arsenal, I think, rebasing off master probably didn't rebase correctly, fixing now

**Comment by rahulreddy** [comment]
@ironman-machine r+

**Comment by rahulreddy** [comment]
@ironman-machine r-

**Comment by rahulreddy** [comment]
@ironman-machine r+

**Comment by rahulreddy** [comment]
@ironman-machine r+

**Comment by JianqinWang** [comment]
@ironman-machine r+

**Comment by DavidPineauScality** [comment]
This shouldn't have been merged by ultron, and was due to me restarting it. I'll remove the changes from the master branch, and reopen a PR.

**Comment by rahulreddy** [comment]
@ironman-machine r+

**Comment by DavidPineauScality** [comment]
after offline discussion with @rahulreddy , we'll keep it merged. hopefully nothing broke CI here.

**Pull request merged**